File: tests/typesafety/typescript/json_parsed_misuse.ts
// expect: fails

declare const penceBrand: unique symbol;
type Pence = number & { readonly [penceBrand]: true };

type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

type Line = { readonly item: string; readonly pence: Pence; readonly quantity: number };

function invoiceTotal(lines: readonly Line[]): Pence {
  return lines.reduce((sum, line) => sum + line.pence * line.quantity, 0) as Pence;
}

export function report(decoded: Json): Pence {
  return invoiceTotal(decoded);
}

File: tests/typesafety/python/json_parsed_misuse.py
# expect: fails
# title: A Json value handed straight to invoice_total, rejected by the checker
# misuse-of: json_parsed
from dataclasses import dataclass
from typing import NewType

type Json = None | bool | int | float | str | list["Json"] | dict[str, "Json"]

Pence = NewType("Pence", int)


@dataclass(frozen=True)
class Line:
    item: str
    pence: Pence
    quantity: int


def invoice_total(lines: list[Line]) -> Pence:
    return Pence(sum(line.pence * line.quantity for line in lines))


def report(decoded: Json) -> Pence:
    return invoice_total(decoded)
